### PR TITLE
JAPI-325 BinaryRecordReader not correctly handle integer overflow for non-standard types

### PR DESCRIPTION
**Pull request by jpmcmu**
- Corrected bug in sign extension code
- Added support for writing int3,int5,int6,int7
- Updated numeric overflow tests for non standard int sizes

Signed-off-by: James McMullan <James.McMullan@lexisnexis.com>

## Type of change:
- [X] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [X] I have created a corresponding JIRA ticket for this submission
- [X] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [X] The change has been fully tested:
  - [X] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java [hunk 1]
```diff
@@ -323,10 +323,51 @@ else if (fd.getDataLen() == 2)
...
+                    this.buffer.put((byte) (((value >> 40) & 0xFF) | signBit));
+                }
+                else if (fd.getDataLen() == 7)
+                {
+                    this.buffer.put((byte) (value & 0xFF));
+                    this.buffer.put((byte) ((value >> 8) & 0xFF));
+                    this.buffer.put((byte) ((value >> 16) & 0xFF));
+                    this.buffer.put((byte) ((value >> 24) & 0xFF));
+                    this.buffer.put((byte) ((value >> 32) & 0xFF));
+                    this.buffer.put((byte) ((value >> 40) & 0xFF));
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java:366]: Everything looks good, I had to find something:
consider generalizing this logic and exposing it in a function we could call for fd.getDataLen() == 3,5,6,7.
The function could be something like:

    public void moveNBytesToBuffer(value, buffer, nBytes)
    {
        for (int i = 0; i <nBytes; i+=8)
        {
        	this.buffer.put((byte) ((value >> i) & 0xFF));
		}
        long signBit = value < 0 ? 0x80L : 0;
        this.buffer.put((byte) (((value >> nBytes*8) & 0xFF) | signBit));
    }

**Review by rpastrana — COMMENTED** [review]
@jpmcmu just one suggestion

**Comment by jpmcmu** [comment]
@rpastrana made code review changes, if it looks good will squash

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java [hunk 2]
```diff
@@ -331,10 +331,21 @@ else if (fd.getDataLen() == 8)
                 {
                     this.buffer.putLong(value);
                 }
+                else if (fd.getDataLen() < 8 && fd.getDataLen() > 0)
+                {
+                    long lastByteIdx = fd.getDataLen() -1;
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java:336]: does this need to be a long?
Actually, I see it's used in the bit masking and shift in line 342

File: dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java [hunk 3]
```diff
@@ -331,10 +331,21 @@ else if (fd.getDataLen() == 8)
...
+                    for (int i = 0; i < lastByteIdx; i++) {
```
Comment (rpastrana) [inline dfsclient/src/main/java/org/hpccsystems/dfs/client/BinaryRecordWriter.java:337]: Minor: open curly to new line.

**Review by rpastrana — APPROVED** [review]
@jpmcmu one minor format issue

**Comment by jpmcmu** [comment]
@rpastrana fixed and squashed

**Pull request merged**